cmd/htcondor-api/token_fetch.go

```diff
@@ -0,0 +1,542 @@
+	}
+
+	scopes := strings.Split(tokenResult.Scope, " ")
+	expiresAt := time.Now().Add(time.Duration(tokenResult.ExpiresIn) * time.Second)
+
+	// Use new refresh token if provided, otherwise keep the old one
+	refreshToken := tokenResult.RefreshToken
+	if refreshToken == "" {
+		refreshToken = tokenInfo.RefreshToken
+	}
+
+	return &TokenInfo{
+		TrustDomain:  tokenInfo.TrustDomain,
+		ClientID:     tokenInfo.ClientID,
+		ClientSecret: tokenInfo.ClientSecret,
+		AccessToken:  tokenResult.AccessToken,
+		RefreshToken: refreshToken,
+		TokenType:    tokenResult.TokenType,
+		ExpiresAt:    expiresAt,
+		Scopes:       scopes,
+		IssuerURL:    tokenInfo.IssuerURL,
+	}, nil
+}
+
+// getTokenInfoPath returns the path to the token info file
+func getTokenInfoPath() string {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		homeDir = "/tmp"
+	}
+	condorDir := filepath.Join(homeDir, ".condor")
+	return filepath.Join(condorDir, "token_info")
+}
+
+// getAccessTokenPath returns the path where access token should be saved
+func getAccessTokenPath(trustDomain string) string {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		homeDir = "/tmp"
+	}
+	tokensDir := filepath.Join(homeDir, ".condor", "tokens.d")
+	// Sanitize trust domain for filename
+	safeDomain := strings.ReplaceAll(trustDomain, ":", "_")
+	safeDomain = strings.ReplaceAll(safeDomain, "/", "_")
+	return filepath.Join(tokensDir, safeDomain)
+}
+
+// loadTokenInfoStore loads token info from disk
+func loadTokenInfoStore(path string) (*TokenInfoStore, error) {
+	//nolint:gosec // Path is from user config directory
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var store TokenInfoStore
+	if err := json.Unmarshal(data, &store); err != nil {
+		return nil, err
+	}
+
+	return &store, nil
+}
+
+// saveTokenInfoStore saves token info to disk
+func saveTokenInfoStore(store *TokenInfoStore, path string) error {
+	// Ensure directory exists
+	dir := filepath.Dir(path)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return err
+	}
+
+	data, err := json.MarshalIndent(store, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(path, data, 0600)
+}
+
+// saveAccessToken saves the access token to tokens.d directory
+func saveAccessToken(tokenInfo *TokenInfo) error {
+	tokensDir := filepath.Dir(getAccessTokenPath(tokenInfo.TrustDomain))
+	if err := os.MkdirAll(tokensDir, 0700); err != nil {
+		return err
+	}
+
+	tokenPath := getAccessTokenPath(tokenInfo.TrustDomain)
+	return os.WriteFile(tokenPath, []byte(tokenInfo.AccessToken), 0600)
+}
+
+// checkExistingToken checks if there's a valid token in tokens.d that won't expire for at least 5 minutes
+func checkExistingToken(trustDomain string) (string, error) {
+	tokenPath := getAccessTokenPath(trustDomain)
+
+	// Check if token file exists
+	//nolint:gosec // G304: Token path is derived from user-provided trust domain, controlled by application logic
+	tokenData, err := os.ReadFile(tokenPath)
```

Comment:
There may be multiple tokens per file; obey the `#` as a comment character.  Iterate through non-empty, non-commented lines to find an appropriate token.